File: lc/coreprep/implementation/patterns_reference_implementation/modified_binary_search/OrderAgnosticBinarySearch.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class OrderAgnosticBinarySearch {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        int k = in.nextInt();

        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
          a[i] = in.nextInt();
        }

        System.out.println(binarySearch(a, k));
      }
    }
  }

  // O(logn) / O(1)
  public static int binarySearch(int[] a, int k) {
    if (a.length == 1) {
      return (a[0] == k ? 0 : -1);
    }

    int low = 0, high = a.length - 1;
    boolean isIncreasing = a[low] <= a[high];

    while (low <= high) {
      int mid = low + (high - low) / 2;

      if (a[mid] == k) {
        return mid;
      } else if (a[mid] < k) {
        if (isIncreasing) {
          low = mid + 1;
        } else {
          high = mid - 1;
        }
      } else {
        if (isIncreasing) {
          high = mid - 1;
        } else {
          low = mid + 1;
        }
      }
    }

    return -1;
  }
}

File: lc/coreprep/implementation/patterns_reference_implementation/modified_binary_search/order_agnostic_binary_search.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

// O(logn) / O(1)
int binary_search(const vector<int> &a, int low, int high, int k) {
  bool is_inc = a[low] < a[high];

  while (low <= high) {
    int mid = low + (high - low) / 2;

    if (a[mid] < k) {
      if (is_inc) {
        low = mid + 1;
      } else {
        high = mid - 1;
      }
    } else if (a[mid] > k) {
      if (is_inc) {
        high = mid - 1;
      } else {
        low = mid + 1;
      }
    } else {
      return mid;
    }
  }

  return -1;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, k;
  cin >> tt;

  while (tt--) {
    cin >> n >> k;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
      cin >> a[i];
    }

    cout << binary_search(a, 0, n - 1, k) << "\n";
  }

  return 0;
}